### PR TITLE
 ChainSyncClient: disconnect from nodes with invalid blocks 

### LINKED ISSUE #466: Chain Sync Client: disconnect from nodes with invalid blocks
The `ChainDB`exposes invalid points through `knownInvalidBlocks :: STM m (Set (Point blk))`. The Chain Sync Client should look at this set and use it to disconnect from upstream nodes that have such blocks in their chains.

### PR DESCRIPTION
**Pull request by mrBliss**
Closes #466.

* New headers are checked when rolling forward
* There is a background thread for each candidate that watches the invalid
  blocks for changes. Whenever it(s fingerprint) changes, it looks in the
  candidate fragment for invalid blocks and disconnects if it found one.

### CODE REVIEW

File: ouroboros-consensus/src/Ouroboros/Consensus/Util/STM.hs [hunk 1]
```diff
@@ -68,6 +84,11 @@ blockUntilJust getMaybeA = do
 blockUntilAllJust :: MonadSTM m => [STM m (Maybe a)] -> STM m [a]
 blockUntilAllJust = mapM blockUntilJust
 
+-- | Simple type that can be used to indicate something in a 'TVar' is
+-- changed.
+newtype Fingerprint = Fingerprint Word64
```
Comment (edsko) [inline ouroboros-consensus/src/Ouroboros/Consensus/Util/STM.hs]: Will this be confusing with fingerprints used in the network layer? (Don't know.)

File: ouroboros-consensus/src/Ouroboros/Consensus/ChainSyncClient.hs [hunk 2]
```diff
@@ -474,6 +501,12 @@ chainSyncClient tracer cfg btime (ClockSkew maxSkew)
     disconnect :: ChainSyncClientException blk -> STM m a
     disconnect = throwM
 
+    -- | Trace any 'ChainSyncClientException' if thrown.
+    traceException :: m a -> m a
```
Comment (edsko) [inline ouroboros-consensus/src/Ouroboros/Consensus/ChainSyncClient.hs]: Wonder if this merits living in `Util`?

File: ouroboros-consensus/src/Ouroboros/Consensus/ChainSyncClient.hs [hunk 3]
```diff
@@ -496,6 +529,44 @@ chainSyncClient tracer cfg btime (ClockSkew maxSkew)
...
+-- fingerprint). Whenever it changes, i.e., a new invalid block is detected,
+-- check whether the current candidate fragment contains any header that is
+-- invalid, if so, disconnect by throwing an 'InvalidBlock' exception.
+--
+-- This function spawns a background thread using the given 'ThreadRegistry'.
+--
+-- The cost of this check is \( O(cand * check) \) where /cand/ is the size of
+-- the candidate fragment and /check/ is the cost of checking whether a block
+-- is invalid (typically \( O(\log(invalid)) \) where /invalid/ is the number
+-- of invalid blocks).
```
Comment (edsko) [inline ouroboros-consensus/src/Ouroboros/Consensus/ChainSyncClient.hs]: Might be worth pointing out here that in the possible though unlikely scenario that the node gives us an invalid block, then rolls back and gives us a valid block, we might not disconnect.

**Comment by mrBliss** [comment]
> LGTM. One thing I was wondering about - in the QSM tests, do we check anything about the fingerprint? If so, requiring that they are equal would seem an overspecification maybe?

We currently don't test `getIsInvalidBlock` directly in the QSM tests. Indirectly we test it through the chain selection, which shouldn't pick invalid blocks. We are not testing the fingerprint in the QSM tests.

File: ouroboros-consensus/src/Ouroboros/Consensus/ChainSyncClient.hs (hunk 2 above)
Comment (mrBliss) [reply ouroboros-consensus/src/Ouroboros/Consensus/ChainSyncClient.hs]: Then the type of the exception would have to be passed via a `Proxy`. I propose that we do this when we want to do this at some other place in the code.

File: ouroboros-consensus/src/Ouroboros/Consensus/Util/STM.hs (hunk 1 above)
Comment (mrBliss) [reply ouroboros-consensus/src/Ouroboros/Consensus/Util/STM.hs]: In the network layer, there's only `FetchStateFingerprint`, so i don't think there's a risk of confusion.

File: ouroboros-consensus/src/Ouroboros/Consensus/ChainSyncClient.hs (hunk 2 above)
Comment (edsko) [reply ouroboros-consensus/src/Ouroboros/Consensus/ChainSyncClient.hs]: At that point we might not remember, but sure, not a big deal either way :)

**Comment by mrBliss** [comment]
uggh

bors r+

**Pull request merged**